.pre-commit-hooks.yaml

```diff
@@ -33,6 +33,12 @@
     entry: '\.warn\('
     language: pygrep
     types: [python]
+-   id: python-private-access-enforcement
+    name: don't access other classes's private members
+    description: 'Allow referencing self._whatever but not something._whatever; disable with #private'
+    entry: (?<!\Wself)(?<![)])[.]_[^_](?!.*#\s*private)
```

Comment:
I think this has too many false positives.  For instance:

```python
@classmethod
def f(cls, foo):
   cls._bar
```

or

```python
class C:
    @staticmethod
    def _f(): ...

    def g(self):
         return C._f()
```

Reply:
Is there a codebase you'd recommend testing against? The ones I'm using might not be complicated enough.

Reply:
Those two examples above highlight the problem already:

```console
$ pre-commit run --files t.py
don't access other classes's private members.............................Failed
hookid: python-private-access-enforcement

t.py:7:        return cls._f()
t.py:10:         return C._f()
```

Reply:
I can add an exception for `cls`. For the second example, wouldn't it be arguably better to write that as:

```python
return self.__class__._f()
```
? I might be able to handle that.

Reply:
FWIW, my reason for thinking this would be useful is that I ran it on two of our codebases, and everything it found was actually bad (outside of tests). Might not be possible to make it apply well to every existing codebase.

Reply:
pylint has W0212 -- maybe try using that?

at least for the official hooks they have to be very low false-positive rate or they reflect poorly on the project / ecosystem :(

Reply:
Oh, OK, that obviously seems like a better solution. I was just pursuing this because our conversation the other day gave me the impression there was no established solution for this.

Reply:
Yeah my B, I thought this would be easier to make without false positives but I want thinking outside the box :S sorry about that. Still useful as a local hook if you can get your repo to adhere to it :+1: